### PR TITLE
Implement SubmitEvent

### LINKED ISSUE #24617: Implement SubmitEvent for form submit events
https://github.com/whatwg/html/pull/4984

### PR DESCRIPTION
**Pull request by pshaughn**
FormSubmitter was already being passed into submit, so there wasn't much work to do here other than making the new event interface itself.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #24617 

- [X] There are tests for these changes

### CODE REVIEW

File: components/script/dom/submitevent.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+#[dom_struct]
+#[allow(non_snake_case)]
+pub struct SubmitEvent {
+    event: Event,
+    submitter: Option<DomRoot<HTMLElement>>,
+}
+
+impl SubmitEvent {
+    pub fn new_inherited(submitter: Option<DomRoot<HTMLElement>>) -> SubmitEvent {
```
Comment (jdm) [inline components/script/dom/submitevent.rs]: No need for this to be public.

File: components/script/dom/submitevent.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        SubmitEvent {
+            event: Event::new_inherited(),
+            submitter: submitter,
+        }
+    }
+
+    pub fn new_initialized(
```
Comment (jdm) [inline components/script/dom/submitevent.rs]: No need for a separate method here; let's inline this code into `new` instead.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/submitevent.rs (hunk 2 above)
Comment (pshaughn) [reply components/script/dom/submitevent.rs]: Right, that's me copy-and-pasting from other event code a little too heavily.

**Comment by pshaughn** [comment]
I took out window-named-getter reliance from a few tests in order to get more meaningful results; they still fail, but now it's for a much more interesting reason and the harness itself doesn't catch an error.

**Comment by jdm** [comment]
@pshaughn Did you forget to push the updated changes?

**Comment by pshaughn** [comment]
Apparently yes. I will get to that!

**Comment by pshaughn** [comment]
OK no the version that changed 4000 files is clearly not the right thing

**Comment by pshaughn** [comment]
I think this is now what I actually meant to push.

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Comment by CYBAI** [comment]
```
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface: existence and properties of interface object
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface object length
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface object name
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface: existence and properties of interface prototype object
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface: existence and properties of interface prototype object's "constructor" property
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface: existence and properties of interface prototype object's @@unscopables property
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?exclude=(Document|Window|HTML.*):
  └ PASS [expected FAIL] SubmitEvent interface: attribute submitter

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface SubmitEvent to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:72:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  │ test@http://web-platform.test:8000/resources/testharness.js:548:30
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1
```

**Comment by CYBAI** [comment]
@bors-servo retry

- #24762

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Comment by jdm** [comment]
@bors-servo retry
* #24726

**Comment by CYBAI** [comment]
@bors-servo retry

- #24726

**Comment by jdm** [comment]
@bors-servo retry
* #24726 #21067

**Pull request merged**